tensorflow/core/kernels/fft_ops.cc

```diff
@@ -97,9 +84,89 @@ class FFTGPUBase : public OpKernel {
   virtual bool IsForward() const = 0;
   virtual bool IsReal() const = 0;
 
- private:
+  // The function that actually computes the FFT
+  virtual void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
+                     Tensor* out) = 0;
+};
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+template <typename Device, typename TInput, typename TOutput,
+          int FFTResultType, int FFTDir, int FFTRank>
+struct FFTFunctor {
+  void operator()(const Device& d,
+                  typename TTypes<TOutput, FFTRank + 1>::Tensor output,
+                  typename TTypes<TInput, FFTRank + 1>::ConstTensor input) {
+      // Cast to non-const data type to ensure we can call the template fft
+      // function
+      // TODO(tillahoffmann): Remove reinterpret_cast once Eigen supports const
+      // arguments for tensor fft
+      auto casted_input =
+       *reinterpret_cast<typename TTypes<TInput, FFTRank + 1>::Tensor*>(&input);
+      // Create the axes (which are always trailing)
+      auto axes = Eigen::ArrayXi::LinSpaced(FFTRank, 1, FFTRank);
+      // Evaluate the fft on the specified device
+      output.device(d) = casted_input.template fft<FFTResultType, FFTDir>(axes);
+  }
+};
+
+template <bool Forward, bool _Real, int FFTRank>
+class FFTCPU : public FFTBase {
+ public:
+  using FFTBase::FFTBase;
+ protected:
+  int Rank() const override { return FFTRank; }
+  bool IsForward() const override { return Forward; }
+  bool IsReal() const override { return _Real; }
+
   void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
-             Tensor* out) {
+             Tensor* out) override {
+    // TODO(tillahoffmann): Support RFFTs by slicing away the negative
+    // frequency components
+    OP_REQUIRES(ctx, !IsReal(), errors::Internal("Real FFT not supported."));
+    auto input = in.flat_inner_dims<complex64, FFTRank + 1>();
+
+    // Apply the functor
+    FFTFunctor<CPUDevice, complex64, complex64, Eigen::BothParts,
+               Forward ? Eigen::FFT_FORWARD : Eigen::FFT_REVERSE,
+               FFTRank> functor;
+    functor(ctx->eigen_device<CPUDevice>(),
+            out->flat_inner_dims<complex64, FFTRank + 1>(), input);
+  }
+};
+
+REGISTER_KERNEL_BUILDER(Name("FFT").Device(DEVICE_CPU), FFTCPU<true, false, 1>);
```

Comment:
Not sure if this would be the right solution, but one way might be to just name these kernels FFT_V2 etc. and have internal Python wrappers calling the old versions while the third_party/tensorflow code would call the new ones?